### PR TITLE
Send email on new lead, sort leads in chart, forms delete request, and style fixes

### PR DESCRIPTION
**Pull request by FluffyPenguin**
# Description

- Now sends an email to the advertiser on new leads. Needs Java 8 for this functionality. Will need to implement a preference for email notifications for the advertiser and needs testing on live server.
- Leads are now able to be sorted by a certain parameter by clicking the respective header at the top of the chart.
- api/forms now supports a DELETE request to delete a certain form entity.
- Style fixes for Java code -> mainly indentation changes from 4 spaces to 2

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected). Requires Java 8 JDK.
- [x] This change requires a documentation update

# How Has This Been Tested?

Manually on the local development server.

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings

### CODE REVIEW

File: backend/src/main/java/com/google/step/data/Form.java [hunk 1]
```diff
@@ -17,77 +17,84 @@
...
 import com.google.appengine.api.datastore.KeyFactory;
-
+import com.google.appengine.api.users.User;
+import com.google.step.utils.AdvertiserUtil;
 import java.util.Date;
 
 /**
- * This class represents a Form and all of its data.
- * Supports conversion to datastore Entity objects and back.
+ * This class represents a Form and all of its data. Supports conversion to datastore Entity objects
```
Comment (anttsou-google) [inline backend/src/main/java/com/google/step/data/Form.java:25]: nit: remove "This class" and just do "Represents". Generally, starting with a verb is preferred.

File: backend/src/main/java/com/google/step/servlets/WebhookServlet.java [hunk 2]
```diff
@@ -67,9 +60,21 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
...
+      response.sendError(404,
+          "Invalid webhook url. Advertiser has not registered with our site.");
+      return;
+    }
     myLead = Lead.fromReader(request.getReader());
     //TODO: Add additional verification steps
+    try {
+      EmailUtil.sendNewLeadEmail(user);
+    } catch (MessagingException e) {
+      System.out.println(e);
```
Comment (anttsou-google) [inline backend/src/main/java/com/google/step/servlets/WebhookServlet.java:76]: What behavior do we want if this fails? Do we want to notify that email failed, or maybe retry (with a limit)?

File: backend/src/main/java/com/google/step/utils/AdvertiserUtil.java [hunk 3]
```diff
@@ -15,62 +15,107 @@
...
+   * Checks whether the user object passed in exists in datastore as an advertiser
+   *
+   * @param user the google user to be checked
+   * @return true if the user is contained in the datastore as an advertiser, false otherwise
+   * @throws DatastoreFailureException if a datastore error occurs
+   */
+  public static boolean advertiserExistsInDatastore(User user) {
+    try {
+      DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+      Key advertiserKey = createAdvertiserKey(user);
```
Comment (anttsou-google) [inline backend/src/main/java/com/google/step/utils/AdvertiserUtil.java:46]: which part throws the EntityNotFoundException? Maybe have that be the only part of the try block.

File: backend/src/main/java/com/google/step/utils/EmailUtil.java [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+import javax.mail.Message;
+import javax.mail.MessagingException;
+import javax.mail.Session;
+import javax.mail.Transport;
+import javax.mail.internet.AddressException;
+import javax.mail.internet.InternetAddress;
+import javax.mail.internet.MimeMessage;
+
+/**
+ * This utility class is responsible for using the App Engine Mail API to handle any emailing that
```
Comment (anttsou-google) [inline backend/src/main/java/com/google/step/utils/EmailUtil.java:30]: Replace with just the verb "Uses", or maybe better "Handles".

File: backend/src/main/java/com/google/step/servlets/WebhookServlet.java (hunk 2 above)
Comment (FluffyPenguin) [reply backend/src/main/java/com/google/step/servlets/WebhookServlet.java]: Hm, I'm not sure what we want to do.
I guess for now, we can log that the email failed and decide whether or not we want to retry?

File: backend/src/main/java/com/google/step/data/Form.java (hunk 1 above)
Comment (FluffyPenguin) [reply backend/src/main/java/com/google/step/data/Form.java]: Fixed. planning to do a quick look over all the javadocs soon.

File: backend/src/main/java/com/google/step/utils/AdvertiserUtil.java (hunk 3 above)
Comment (FluffyPenguin) [reply backend/src/main/java/com/google/step/utils/AdvertiserUtil.java]: Moved the datastore and advertiserKey lines out of the try block

File: backend/src/main/java/com/google/step/utils/EmailUtil.java (hunk 4 above)
Comment (FluffyPenguin) [reply backend/src/main/java/com/google/step/utils/EmailUtil.java]: Fixed

**Pull request merged**